src/node/filesystem/binary.node.ts

```diff
@@ -0,0 +1,341 @@
+    this.buffer = this.buffer.subarray(size);
+
+    return chunk;
+  }
+}
+
+export class ReadableClone extends Readable {
+  constructor(private stream: Readable, options?: ReadableOptions) {
+    super(options);
+
+    this.stream.on('data', (chunk) => {
+      this.push(chunk);
+    });
+
+    this.stream.on('end', () => {
+      this.push(null);
+    });
+
+    this.stream.on('error', (error) => {
+      this.emit('error', error);
+    })
+  }
+
+  public override _read(): void {}
+}
+
+export class FileContainer implements IDataContainer, IBinaryFileMeta, IInitializable, IDestructible {
+  private handle: FileHandle | undefined;
+  private stream: Readable | undefined;
+  private streamReader: StreamReader | undefined;
+  public filesize = Infinity;
+  public filename: string | undefined;
+  public mimetype: string | undefined;
+
+  constructor(public path: string, options: { filename?: string, mimetype?: string } = {}) {
+    this.mimetype = options.mimetype;
+    this.filename = options.filename;
+  }
+
+  public async read(size?: number): Promise<Buffer | undefined> {
+    if (!this.streamReader) {
+      throw new UnexpectedError('File not initialized');
+    }
+
+    return await this.streamReader.read(size);
+  }
+
+  public async initialize(): Promise<void> {
+    if (this.handle === undefined) {
+      try {
+        this.handle = await open(this.path, 'r');
+        const { size } = await this.handle.stat();
+        this.filesize = size;
+      } catch (error) {
+        throw handleNodeError(error);
+      }
+    }
+
+    if (this.handle === undefined) {
+      throw new UnexpectedError('Unable to initialize file');
+    }
+
+    let stream: Readable;
+    // We need to create a stream the old fashioned way for Node < 16.11.0
+    if (typeof this.handle.createReadStream !== 'function') {
+      stream = createReadStream(this.path);
+    } else {
+      stream = this.handle.createReadStream();
+    }
+
+    this.stream = new ReadableClone(stream);
+    this.streamReader = new StreamReader(new ReadableClone(stream));
+  }
+
+  public async destroy(): Promise<void> {
+    if (this.handle !== undefined) {
+      try {
+        await this.handle.close();
+        this.stream = undefined;
+        this.streamReader = undefined;
+        this.filesize = Infinity;
+        this.handle = undefined;
+      } catch (_) {
+        // Ignore
+      }
+    }
+  }
+
+  public toStream(): Readable {
+    if (this.stream === undefined) {
+      throw new UnexpectedError('File not initialized');
+    }
+
+    return this.stream;
+  }
+}
+
+export class StreamContainer implements IDataContainer {
+  private stream: Readable;
+  private streamReader: StreamReader;
+
+  constructor(stream: Readable) {
+    this.stream = new ReadableClone(stream);
+    this.streamReader = new StreamReader(new ReadableClone(stream));
+  }
+
+  public read(size?: number): Promise<Buffer | undefined> {
+    return this.streamReader?.read(size);
```

Comment:
nitpick: probably doesn't need a `?` anymore?

Reply:
You are right. It will be oversight leftover from quick prototyping. Thanks